nms/app/packages/magmalte/app/views/equipment/GatewayDetailMain.js

```diff
@@ -73,43 +85,147 @@ const useStyles = makeStyles(theme => ({
   },
 }));
 
-function GatewayRebootButtonInternal(props: WithAlert) {
+const StyledMenu = withStyles({
+  paper: {
+    border: '1px solid #d3d4d5',
+  },
+})(props => (
+  <Menu
+    data-testid="policy_menu"
+    elevation={0}
+    getContentAnchorEl={null}
+    anchorOrigin={{
+      vertical: 'bottom',
+      horizontal: 'center',
+    }}
+    transformOrigin={{
+      vertical: 'top',
+      horizontal: 'center',
+    }}
+    {...props}
+  />
+));
+
+type CommandProps = {
+  gatewayID: string,
+  open: boolean,
+  onClose: () => void,
+};
+
+function GatewayCommandDialog(props: CommandProps) {
+  return (
+    <Dialog open={props.open} onClose={props.onClose} scroll="body">
+      <DialogTitle onClose={props.onClose} label={'Run Gateway Commands'} />
+      <DialogContent>
+        <PingCommandControls gatewayID={props.gatewayID} />
+        <GenericCommandControls gatewayID={props.gatewayID} />
+      </DialogContent>
+    </Dialog>
+  );
+}
+
+function GatewayMenuInternal(props: WithAlert) {
   const classes = useStyles();
   const {match} = useRouter();
   const networkId: string = nullthrows(match.params.networkId);
   const gatewayId: string = nullthrows(match.params.gatewayId);
+  const [anchorEl, setAnchorEl] = React.useState(null);
+  const [open, setOpen] = React.useState(false);
   const enqueueSnackbar = useEnqueueSnackbar();
 
-  const handleClick = () => {
-    props
-      .confirm(`Are you sure you want to reboot ${gatewayId}?`)
-      .then(async confirmed => {
-        if (!confirmed) {
-          return;
-        }
-        try {
-          await RunGatewayCommands({networkId, gatewayId, command: 'reboot'});
-          enqueueSnackbar('gateway reboot triggered successfully', {
-            variant: 'success',
-          });
-        } catch (e) {
-          enqueueSnackbar(e.response?.data?.message ?? e.message, {
-            variant: 'error',
-          });
-        }
-      });
+  const handleClick = event => {
+    setAnchorEl(event.currentTarget);
+  };
+
+  const handleClose = () => {
+    setAnchorEl(null);
   };
 
   return (
-    <Button
-      variant="contained"
-      className={classes.appBarBtn}
-      onClick={handleClick}>
-      Reboot
-    </Button>
+    <div>
+      <GatewayCommandDialog
+        gatewayID={gatewayId}
+        open={open}
+        onClose={() => setOpen(false)}
+      />
+      <Button
+        onClick={handleClick}
+        className={classes.appBarBtn}
+        endIcon={<ArrowDropDownIcon />}>
+        Actions
+      </Button>
+      <StyledMenu
+        anchorEl={anchorEl}
+        keepMounted
+        open={Boolean(anchorEl)}
+        onClose={handleClose}>
+        <MenuItem
+          data-testid="gatewayReboot"
+          onClick={() => {
```

Comment:
this and the `onClick` below seem long enough to warrant pulling out. You can also probably make it generalized and reuse.